src/Control/Effect/Interpose.hs

```diff
@@ -36,19 +36,19 @@ interpose m f = send (Interpose m f pure)
 runInterpose :: InterposeC eff m a -> m a
 runInterpose = runReader Nothing . runInterposeC
 
-newtype InterposeC eff m a = InterposeC { runInterposeC :: ReaderC (Maybe (Listener eff (InterposeC eff m))) m a }
-  deriving (Alternative, Applicative, Functor, Monad)
+newtype InterposeC (eff :: (* -> *) -> * -> *) m a = InterposeC
+  { runInterposeC :: ReaderC (Maybe (Listener eff (InterposeC eff m))) m a
+  } deriving (Alternative, Applicative, Functor, Monad)
 
-newtype Listener eff m = Listener (forall n x . eff n (n x) -> m x)
+newtype Listener (eff :: (* -> *) -> * -> *) m = Listener (forall n x . eff n x -> m x)
 
--- TODO: Document the implementation of this, as it is extremely subtle.
-
-runListener :: Listener eff (InterposeC eff m) -> eff (InterposeC eff m) (InterposeC eff m a) -> InterposeC eff m a
+-- Normally we can't just extract the existentials out of the Listener type. In this case,
+-- we can constrain the foralled 'n' variable to be 'Interpose', which lets it by the typechecker.
```

Comment:
I took a shot at explaining this, but @robrix might have a better explanation for how it does what it does.

Reply:
I have zero recollection of how any of this ever worked, tbh.